test/models/product_test.rb

```diff
@@ -0,0 +1,200 @@
+        invalid_product.errors.messages.must_include :price
+      end
+    end
+
+    it "requires a name" do
+      product = Product.new(category: 'food', price: 10.00, quantity: 5, user: tanja)
+      product.valid?.must_equal false
+      product.errors.messages.must_include :name
+    end
+
+    it "requires unique names w/in categories" do
+      category = 'food'
+      name = 'test name'
+      product1 = Product.new(name: name, category: category, price: 10.00, quantity: 5, user: tanja)
+      product1.save!
+
+      product2 = Product.new(name: name, category: category, price: 10.00, quantity: 5, user: tanja)
+      product2.valid?.must_equal false
+      # product2.errors.messages.must_include :name
+    end
+
+    it "does not require a unique name if the category is different" do
+      name = 'test name'
+      product1 = Product.new(name: name, category: 'food', price: 10.00, quantity: 5, user: tanja)
+      product1.save!
+
+      product2 = Product.new(name: name, category: 'cosmetics', price: 10.00, quantity: 5, user: tanja)
+      product2.valid?.must_equal true
+    end
+  end
+
+  describe "custom model methods " do
+
+    describe "self.by_category(category)" do
+      let(:transportation) { categories(:transportation) }
+      let(:weekend_yacht) {products(:yacht)}
+      let(:cosmetics_cat) { categories(:cosmetics)}
+      it "returns a collection of Products of the appropriate category" do
+        weekend_yacht.categories << transportation
+        Product.by_category(transportation).count.must_equal 1
+        Product.by_category(transportation)[0].must_be_kind_of Product
+
+        ###REFACTOR tests below ####
+        cosmetics = Product.where(category: "cosmetics")
+
+        cosmetics.each do |cosmetic|
+          cosmetic.categories << cosmetics_cat
+        end
+
+        Product.by_category(cosmetics_cat).must_be_kind_of Enumerable
+
+        Product.by_category(cosmetics_cat).count.must_equal 3
+
+      end
+
+      it "will not error out if there are no products of that category" do
+
+        Product.by_category("bamboozles").count.must_equal 0
+      end
+
+    end
+
+    describe "self.by_merchant(merchant)" do
+      let(:user) {users(:bubbles)}
+      it "returns an array of products by the given merchant" do
+        Product.by_merchant(user).count.must_equal user.products.count
+
+        user.products.each do |product|
+          Product.by_merchant(user).must_include product
+        end
+
+        Product.by_merchant(users(:buttercup)).count.must_equal 2
+
+      end
+
+      it "will not error out if there are no products from that merchant" do
+        Product.by_category("bamboozled").count.must_equal 0
+      end
+
+    end
+
+    describe "self.to_merchant_hash method" do
+      it "returns a hash with user names as keys and product arrays as values" do
+        merchant_hash = Product.to_merchant_hash
+
+        merchant_hash.must_be_kind_of Hash
+        merchant_hash.keys.count.must_equal 4
```

Comment:
I would instead loop through an array of the keys and say:

`merchant_hash.keys.must_include key`